lib/livebook/file_system.ex

```diff
@@ -0,0 +1,133 @@
+defprotocol Livebook.FileSystem do
+  @moduledoc false
+
+  # This protocol defines an interface for file systems
+  # that can be plugged into Livebook.
+
+  @typedoc """
+  A path uniquely idenfies file in the file system.
+
+  Path has most of the semantics of regular file paths,
+  with the following exceptions:
+
+    * path must be be absolute for consistency
+
+    * directory path must have a trailing slash, whereas
+      regular file path must not have a trailing slash.
+      Rationale: some file systems allow a directory and
+      a file with the same name to co-exist, while path
+      needs to distinguish between them
+  """
+  @type path :: String.t()
+
+  @typedoc """
+  A human-readable error message clarifying the operation
+  failure reason.
+  """
+  @type error :: String.t()
+
+  @type access :: :read | :write | :read_write | :none
+
+  @doc """
+  Returns the default directory path.
+
+  To some extent this is similar to current working directory
+  in a regular file system. For most file systems this
+  will just be the root path.
+  """
+  @spec default_path(t()) :: path()
+  def default_path(file_system)
+
+  @doc """
+  Returns a list of files located in the given directory.
+  """
+  @spec list(t(), path(), boolean()) :: {:ok, list(path())} | {:error, error()}
+  def list(file_system, path, recursive)
+
+  @doc """
+  Returns binary content of the given file.
+  """
+  @spec read(t(), path()) :: {:ok, binary()} | {:error, error()}
+  def read(file_system, path)
+
+  @doc """
+  Writes the given binary content to the given file.
+
+  If the file exists, it gets overridden.
+
+  If the file doesn't exist, it gets created along with
+  all the necessary directories.
+  """
+  @spec write(t(), path(), binary()) :: :ok | {:error, error()}
+  def write(file_system, path, content)
+
+  @doc """
+  Returns the current access level to the given file.
+
+  If determining the access is costly, then this function may
+  always return the most liberal access, since all access
+  functions return error on an invalid attempt.
+  """
+  @spec access(t(), path()) :: {:ok, access()} | {:error, error()}
```

Comment:
Hrm, what do we need access for? Would it be easier if we instead change the semantics of `list` to not list anything that we may not have access to?

Reply:
There may be read-only notebooks, in which case we disable the "Open" button with a tooltip, but it's still forkable. I think the nerves built-in notebooks are like that.

Reply:
For S3 it always returns `:read_write`, so that we don't do an unnecessary request.